lib/Frontend/DependencyVerifier.cpp

```diff
@@ -0,0 +1,680 @@
+//===--- DependencyVerifier.cpp - Dependency Verifier ---------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  Implements a verifier for dependencies registered against the
+//  ReferencedNameTracker in a SourceFile.
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/AST/ASTMangler.h"
+#include "swift/AST/ASTPrinter.h"
+#include "swift/AST/SourceFile.h"
+#include "swift/Basic/OptionSet.h"
+#include "swift/Frontend/DiagnosticVerifier.h"
+#include "swift/Parse/Lexer.h"
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/Support/FormatVariadic.h"
+
+using namespace swift;
+
+namespace {
+
+/// An \c Expectation represents a user-provided expectation for a particular
+/// dependency entry. An expectation is usually written in-line in a comment
+/// attached near the relevant declaration and takes one of the following forms:
+///
+/// // expected-provides {{ProvidedName}}
+/// // expected-private-member {{some.User.member}}
+///
+/// An expectation contains additional information about the expectation
+/// \c Kind, which matches one of the few kinds of dependency entry that are
+/// currently representable in the dependency graph, and an expectation
+/// \c Flavor which must either be \c private or \c cascading.
```

Comment:
I designed this tool in intentional ignorance of the driver's view of dependencies and dependency graphs. It is meant to test the `ReferencedNameTracker`, and I believe it should continue to follow its terminology until such time as we obviate the implementation. Now, I will integrate the change from `Flavor` to `Scope`. You picked up on the fact that I was fumbling for a good noun there. Thanks!

Reply:
I agree that compartmentalization would push this tool towards its own terminology, but
I have to disagree with the notion that this tool ought to be ignorant of the driver's view of dependencies. The whole purpose of this tool is to ensure that the driver does the right thing when scheduling jobs for incremental recompilation. Thus, there are many invariants that must span the frontend and driver, some of which are being checked by this tool. Since the invariants span, why not the terminology?

What am I missing here? Would you care to expand on your thinking?

Reply:
> The whole purpose of this tool is to ensure that the driver does the right thing when scheduling jobs for incremental recompilation

That is not its goal. The verification machinery you wrote should be testing that. What this tool aims to surface are bugs caused by ordering dependencies that result from the way jobs are scheduled, certainly. But the goal is debug the frontends in such a case, not the driver.